feat(cluster-agent): adding cluster-agent with node-agent registration

mynktl opened the pull request:
changes:
- Added initial code for cluster-agent to support registration of node-agent.
- Updated deployer tool to run cluster-agent
- Added bdd for cluster-agent

Signed-off-by: Mayank <mayank.patel@datacore.com>

tiagolobocastro commented on control-plane/agents/Cargo.toml:
```diff
@@ -21,6 +21,10 @@ path = "common/src/lib.rs"
  path = "ha-node/src/main.rs"
 
+[[bin]]
+ name = "cluster-agent"
```
agent-cluster

tiagolobocastro commented on control-plane/agents/Cargo.toml:
```diff
@@ -21,6 +21,10 @@ path = "common/src/lib.rs"
 
+[[bin]]
+ name = "cluster-agent"
+ path = "cluster-agent/src/main.rs"
```
we're already in agents, so cluster would suffice

tiagolobocastro commented on control-plane/agents/cluster-agent/src/main.rs:
```diff
@@ -0,0 +1,29 @@
+struct Cli {
+    /// IP address and port for the cluster-agent to listen on
+    #[structopt(long, short, default_value = DEFAULT_CLUSTER_AGENT_SERVER_ADDR)]
+    grpc_endpoint: Uri,
```
Should be `SocketAddr` ?

Commit:
feat(deployer): adding support for HA agent-cluster

Signed-off-by: Mayank <mayank.patel@datacore.com>

tiagolobocastro commented on common/src/types/v0/transport/cluster_agent.rs:
```diff
@@ -0,0 +1,18 @@
+#[derive(Debug)]
+pub struct NodeAgentInfo {
+    /// node name where node-agent is running
+    pub node_name: String,
```
instead of making the internal pub, can we add accessors, example:
`pub fn node_name(&self) -> &str { &self.node }`

tiagolobocastro commented on common/src/types/v0/transport/cluster_agent.rs:
```diff
@@ -0,0 +1,18 @@
+    /// node name where node-agent is running
+    pub node_name: String,
+    /// endpoint of node-agent GRPC server
+    pub endpoint: String,
```
should this be a `SocketAddr` or `Url` ?

tiagolobocastro commented on control-plane/agents/cluster/src/server.rs:
```diff
@@ -0,0 +1,51 @@
+            ));
+        }
+
+        tracing::trace!("node agent={} successfully registered", request.node());
```
```suggestion
        tracing::trace!(agent=request.node(), "node successfully registered");
```

tiagolobocastro commented on control-plane/agents/cluster/src/server.rs:
```diff
@@ -0,0 +1,51 @@
+use std::{net::SocketAddr, sync::Arc};
+use tonic::transport::Server;
+
+pub struct ClusterAgent {
```
These don't need to be pub?

mynktl replied:
no, it is being used in control-plane/agents/cluster/src/main.rs

mynktl replied:
no. It is used in [main.rs](https://github.com/openebs/mayastor-control-plane/blob/abad24343a59980563fb8efb751a8c77aa4f3dfa/control-plane/agents/cluster/src/main.rs#L26)

tiagolobocastro replied:
They still don't need to be pub, they can be pub(crate).

tiagolobocastro commented on tests/bdd/features/ha/cluster-agent/cluster_agent.feature:
```diff
@@ -0,0 +1,15 @@
+
+
+Scenario: register node-agent
+    When RegisterNodeAgent request is sent to cluster-agent
```
the RegisterNodeAgent

tiagolobocastro commented on tests/bdd/features/ha/cluster-agent/cluster_agent.feature:
```diff
@@ -0,0 +1,15 @@
+
+Scenario: register node-agent
+    When RegisterNodeAgent request is sent to cluster-agent
+    Then request should be completed successfully
```
the request

tiagolobocastro commented on control-plane/agents/ha-node/src/main.rs:
```diff
@@ -47,7 +48,15 @@ async fn main() {
-        .register(cli.node_name.clone(), cli.grpc_endpoint)
+        .register(&NodeAgentInfo::new(
+            cli.node_name.clone(),
+            cli.grpc_endpoint
```
grpc_endpoint should be a `SocketAddr`? Then we don't need to convert here.

mynktl replied:
yes. I've kept it as Url to keep it consistent with other binaries.

Commit:
feat(agent-cluster): adding intial code for HA cluster-agent

This commit adds initial code for cluster-agent supporting registration for node-agent.
For now, registration doesn't store information about node-agent.

Signed-off-by: Mayank <mayank.patel@datacore.com>

mynktl commented:
bors try

tiagolobocastro commented:
bors delegate
bors try

tiagolobocastro commented:
bors delegate

tiagolobocastro commented:
bors d+

mynktl commented:
bors merge
